include/pybind11/detail/function_record_pyobject.h

```diff
@@ -0,0 +1,205 @@
+    /* PyObject *tp_mro */ nullptr,
+    /* PyObject *tp_cache */ nullptr,
+    /* PyObject *tp_subclasses */ nullptr,
+    /* PyObject *tp_weaklist */ nullptr,
+    /* destructor tp_del */ nullptr,
+    /* unsigned int tp_version_tag */ 0,
+    /* destructor tp_finalize */ nullptr,
+#if PY_VERSION_HEX >= 0x03080000
+    /* vectorcallfunc tp_vectorcall */ nullptr,
+#endif
+};
+PYBIND11_WARNING_POP
+
+static bool function_record_PyTypeObject_PyType_Ready_first_call = true;
+
+inline void function_record_PyTypeObject_PyType_Ready() {
+    if (function_record_PyTypeObject_PyType_Ready_first_call) {
+        if (PyType_Ready(&function_record_PyTypeObject) < 0) {
+            throw error_already_set();
+        }
+        function_record_PyTypeObject_PyType_Ready_first_call = false;
+    }
+}
+
+inline bool is_function_record_PyObject(PyObject *obj) {
+    if (PyType_Check(obj) != 0) {
+        return false;
+    }
+    PyTypeObject *obj_type = Py_TYPE(obj);
+    // Fast path (pointer comparison).
+    if (obj_type == &function_record_PyTypeObject) {
+        return true;
+    }
+    // This works across extension modules. Note that tp_name is versioned.
+    if (strcmp(obj_type->tp_name, function_record_PyTypeObject.tp_name) == 0) {
+        return true;
+    }
+    return false;
+}
+
+inline function_record *function_record_ptr_from_PyObject(PyObject *obj) {
+    if (is_function_record_PyObject(obj)) {
+        return ((detail::function_record_PyObject *) obj)->cpp_func_rec;
+    }
+    return nullptr;
+}
+
+inline object function_record_PyObject_New() {
+    auto *py_func_rec = PyObject_New(function_record_PyObject, &function_record_PyTypeObject);
+    if (py_func_rec == nullptr) {
+        throw error_already_set();
+    }
+    py_func_rec->cpp_func_rec = nullptr; // For clarity/purity. Redundant in practice.
+    return reinterpret_steal<object>((PyObject *) py_func_rec);
+}
+
+PYBIND11_NAMESPACE_BEGIN(function_record_PyTypeObject_methods)
+
+// Guard against accidents & oversights, in particular when porting to future Python versions.
+inline PyObject *tp_new_impl(PyTypeObject *, PyObject *, PyObject *) {
+    pybind11_fail("UNEXPECTED CALL OF function_record_PyTypeObject_methods::tp_new_impl");
+    // return nullptr; // Unreachable.
+}
+
+inline PyObject *tp_alloc_impl(PyTypeObject *, Py_ssize_t) {
+    pybind11_fail("UNEXPECTED CALL OF function_record_PyTypeObject_methods::tp_alloc_impl");
+    // return nullptr; // Unreachable.
+}
+
+inline int tp_init_impl(PyObject *, PyObject *, PyObject *) {
+    pybind11_fail("UNEXPECTED CALL OF function_record_PyTypeObject_methods::tp_init_impl");
+    // return -1; // Unreachable.
+}
+
+// The implementation needs the definition of `class cpp_function`.
+void tp_dealloc_impl(PyObject *self);
+
+inline void tp_free_impl(void *) {
+    pybind11_fail("UNEXPECTED CALL OF function_record_PyTypeObject_methods::tp_free_impl");
+}
+
+inline PyObject *reduce_ex_impl(PyObject *self, PyObject *, PyObject *) {
```

Comment:
I worry this implementation violated the protocol: the returned function call should have created an object of the same type as the one that was reduce_ex-d. aka:

```
reduce = obj.__reduce_ex__()
type(reduce[0](*reduce[1])) == type(obj)
```

Reply:
Quoting from the Python documentation:

* https://docs.python.org/3/library/pickle.html#object.__reduce__

> * A callable object that will be called to create the initial version of the object.
> * A tuple of arguments for the callable object. An empty tuple must be given if the callable does not accept any argument.

There is no mention that the callable has to be of the same type is the object being pickled. Such a restriction would be completely artificial and severely limiting:

There are types that for good reasons are not referenced from any importable module. There is no way that `pickle.load()` could find them.

This is the case for Boost.Python and pybind11 functions.

Going deeper into the weeds:

Python has `__builtins__` as home for some similar types. I was thinking of adding the pybind11 function type to `__builtins__`, but I was told that future Python versions will make this impossible.

Even deeper:

We could create something like `__pybind11__builtins__` (I'd actually call that `__pybind11_internals__`), which would have the benefit of solving problems like [here](https://github.com/pybind/cmake_example/pull/164#issuecomment-1821387592), BUT, how does that spring into existence? — Python creates `__builtins__` on startup. We don't have that luxury.

Also note that there may be multiple pybind11 versions / ABIs in use simultaneously in any given Python interpreter.

Conclusion: I don't think there is any practical way to implement a callable of the "correct" pybind11 function record type that `pickle.load()` could find.

Reply:
My main concern is the type of the return value of the callable, which is supposed to be the "initial version of the object". From what I read the current approach is:

1. function_recorder's callable returns a module object, even though there is a PyObject for function_record itself.
2. Another Python side object's "callable" looks up the attr from the module object to return the callable. (I could not quite find the relevant implementation however).

I argue that step 1's callable should have returned the PyObject for function_record instead, because of the "returns initial value" requirement.  As function_record has a name, and a scope, I think we should have enough information for that. The records can be either attached to a pybind11_builtin module, or the actual scope where the CCallable is attached.

Then step 2 can be adjusted accordingly.

Reply:
"Initial" here refers to what happens when the object is unpickled.

When the object is unpickled: the callable creates an object, then calls `__getstate__` (if there is one) to mutate the "initial" object.

________


**The step 2. is something we have no control over.**

Because it is coded into the pickle implementation how it handles `built-in method` objects `repr(m.simple_callable)`:

```
<built-in method simple_callable of pybind11_detail_function_record_v1__gcc_libstdcpp_cxxabi1018 object at 0x...>
```

Unless we want to completely change how pybind11 functions are implemented (nanobind actually did that!), we are forced to design the `__reduce_ex__` of the function_record to produce:

A callable that returns an (intermediate) object for with `getattr(obj, 'simple_callable`)` returns our function object.

The intermediate object could be anything that has that behavior.

Additional important requirement: it needs to assume that the module with the function was not imported already, otherwise the pickle load step (we have to assume it's in a new process) will fail.

In other words, in or before the `getattr(obj, 'simple_callable')` step the module needs to be imported, and we cannot rely on any side-effects of importing the module before we trigger the import:

Yes, the function object has a name and a scope, but the type object does not have a scope.

This will be a problem even for nanobind, which doesn't use `PyCFunction` but has it's own `nanobind.nb_func` type. To stay with that concrete example: unless `nanobind` exists as an importable module, the pickle load stage cannot work. This is what I meant by the "spring into existence" problem mentioned before.

I was thinking of giving giving the function_record type an `__init__` that could act as the callable in the pickle mechanism, but that's the exact same problem: how does it spring into existence in the pickle load stage?

Note that this PR solves that problem in a very minimalistic and clean fashion:

```
(<built-in function eval>, ("__import__('importlib').import_module('pybind11_tests.pickling')",))
```

It could be even better (avoid the `eval`) if we had something like `__import_module__`, which could then act as the callable directly:

```
(<built-in function __import_module__>, ("pybind11_tests.pickling",))
```

Unfortunately that does not exist. I actually tried under this PR to make such a callable myself, and to inject it into the `rec->scope` when a `cpp_function` is built (so that it springs into existence just in time), but that fell flat because stubgen picks it up (>500 TGP failures because we have ~130 auto-generated but checked-in stubgen files). See message of the commit that replaced the "pickle helper" callable with `eval`:

0e7bb1034e3acfbde139594ebcb84d37b4dfc023

I had `_function_record_pickle_helper_abi_stamp` as the name at that time. I also tried `__function_record_pickle_helper_abi_stamp` and `__function_record_pickle_helper_abi_stamp__` but stubgen stubbornly picked it up no matter what. The `eval` approach was born out of an intense search for a solution that does not make it necessary to change stubgen, or worse, 130+ checked-in auto-generated files all over the codebase.

Reply:
Thanks a lot @rainwoodman for taking the time to go through this yesterday 1:1, and for correcting my understanding of your concern.

I captured the experimental code we worked on in 02b31b1b75979ff36985127056937cd3c3061f2d (exactly as it was when we left off).

Our conclusion is captured in the long comment added with bcdfc6980169dabb7848122602158aeffe41f423. Could you please take another look?

Reply:
Thanks for adding the comment and the test. This looks good!